### PR TITLE
Mbliakharskyi/asana

### PR DESCRIPTION
**Pull request by mbliakharskyi**
## Description

Please include a summary of the changes and issues addressed in this PR. Also, explain the motivation and context of each update. List any dependencies or prerequisites needed for implementing this change.

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/asana/src/connectors/users.ts [hunk 1]
```diff
@@ -31,19 +31,20 @@ export const getUsers = async (accessToken: string, workspace: string, offset?:
...
-      workspace,
-      limit: env.USERS_SYNC_JOB_BATCH_SIZE,
-      offset: offset,
-      opt_fields:
-        'email, name,resource_type, offset, path,uri,workspaces,workspaces.name,workspaces.resource_type, role',
-    });
-    /* eslint-enable -- no type here */
+    const opt_fields = 'email, name';
+    
+    const response = await fetch(`https://app.asana.com/api/1.0/users?opt_fields=${opt_fields}&workspace=${workspace}&limit=${env.USERS_SYNC_JOB_BATCH_SIZE}&${offset ? offset : ""}`, {
```
Comment (elbalexandre) [inline apps/asana/src/connectors/users.ts:36]: nice to get ride of the asana sdk 👍

File: apps/asana/src/connectors/webhook.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ export const registerWebhook = async (organisationId: string, accessToken: strin
     Asana.ApiClient.instance.authentications.token = accessToken;
```
Comment (elbalexandre) [inline apps/asana/src/connectors/webhook.ts:9]: We can remove this file as it's unused.

File: apps/asana/src/inngest/functions/users/refresh-token.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import { Organisation } from '@/database/schema';
+import { inngest } from '../../client';
+
+export const tokenRefresh = inngest.createFunction(
+  { id: 'refresh-token' },
+  { event: "token/refresh" },
+  async ({ event, step }) => {
+    const { organisationId, refreshTokenInfo } = event.data;
+
+    const getRefreshToken = await step.run('get-refresh-token',async () => {
```
Comment (elbalexandre) [inline apps/asana/src/inngest/functions/users/refresh-token.ts:14]: We don't need a step when there is only a single step.

File: apps/asana/src/inngest/functions/users/schedule-token-refresh.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import { db } from '@/database/client';
+import { Organisation } from '@/database/schema';
+import { inngest } from '../../client';
+
+export const scheduleTokenRefresh = inngest.createFunction(
+  { id: 'schedule-token-refresh' },
+  { cron: env.TOKEN_REFRESH_CRON },
+  async ({ step }) => {
+
+    const thirtyMinutesAgo = new Date(Date.now() - 30 * 60000);
```
Comment (elbalexandre) [inline apps/asana/src/inngest/functions/users/schedule-token-refresh.ts:12]: We should select token that will expires in thirty minutes
```suggestion
    const thirtyMinutesFromNow = new Date(Date.now() + 30 * 60000);
```

File: apps/asana/src/inngest/functions/users/schedule-token-refresh.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    const organisations = await db
+        .select({
+            id: Organisation.id,
+            refreshToken: Organisation.refreshToken,
+        })
+        .from(Organisation)
+        .where(gte(Organisation.expiresAt, thirtyMinutesAgo));
```
Comment (elbalexandre) [inline apps/asana/src/inngest/functions/users/schedule-token-refresh.ts:20]:
```suggestion
        .where(lte(Organisation.expiresAt, thirtyMinutesFromNow));
```

File: apps/asana/package.json [hunk 6]
```diff
@@ -27,6 +27,7 @@
     "inngest": "^3.6.2",
     "msw": "^2.0.1",
     "next": "14.0.1",
+    "querystring": "^0.2.1",
```
Comment (elbalexandre) [inline apps/asana/package.json:30]: Looks like it's unused
```suggestion
```

File: apps/asana/src/inngest/client.ts [hunk 7]
```diff
@@ -13,6 +13,14 @@ export const inngest = new Inngest({
         offset?: string;
       };
     };
+    'token/refresh': {
+      data: {
+        organisationId: string;
+        refreshTokenInfo: string;
+        syncStartedAt: number;
+        isFirstSync: boolean;
```
Comment (elbalexandre) [inline apps/asana/src/inngest/client.ts:21]: We don't need this property for token refreshing
```suggestion
```

File: apps/asana/src/inngest/functions/users/refresh-token.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable camelcase -- conveniency */
+import { eq } from 'drizzle-orm';
+import { refreshToken, type RefreshTokenResponseData } from '@/connectors/auth';
+import { db } from '@/database/client';
+import { Organisation } from '@/database/schema';
+import { inngest } from '../../client';
+
+export const tokenRefresh = inngest.createFunction(
+  { id: 'refresh-token' },
```
Comment (elbalexandre) [inline apps/asana/src/inngest/functions/users/refresh-token.ts:9]: let's add some seatbelt here:
```suggestion
  {
    id: 'refresh-token',
    concurrency: {
      key: 'event.data.organisationId',
      limit: 1,
    },
    retries: 3,
  },
```

File: apps/asana/src/inngest/functions/users/refresh-token.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable camelcase -- conveniency */
+import { eq } from 'drizzle-orm';
+import { refreshToken, type RefreshTokenResponseData } from '@/connectors/auth';
+import { db } from '@/database/client';
+import { Organisation } from '@/database/schema';
+import { inngest } from '../../client';
+
+export const tokenRefresh = inngest.createFunction(
```
Comment (elbalexandre) [inline apps/asana/src/inngest/functions/users/refresh-token.ts:8]: could be named `refreshToken`

**Review by elbalexandre — COMMENTED** [review]
Nice work ! Beside some minor points it's looks great.

Could you add some test for the inngest function you had ?
You will probably need to mock elba sdk, you can find [here](https://github.com/elba-security/elba-security/blob/6587d8ab29aaed0085255ee6f2a234331c619f76/apps/github/src/__mocks__/elba-sdk.ts) an example of mock, its usage is [here](https://github.com/elba-security/elba-security/blob/6587d8ab29aaed0085255ee6f2a234331c619f76/apps/github/src/inngest/functions/users/sync-users-page.test.ts). Note that we will probably provide an util for that in the futur.

File: apps/asana/src/inngest/functions/users/refresh-token.ts (hunk 9 above)
Comment (mbliakharskyi) [reply apps/asana/src/inngest/functions/users/refresh-token.ts:8]: Already added refreshToken function to /src/connectors/auth

**Pull request closed without merging**